Make Residents act as list

zbarnard1 opened the pull request:
Adds a new dependency that will help us track the position of the Residents for the User to pray for, as well as 2 endpoints for retrieving either a particular Resident by ID or the next resident the User will pray for based on the last Prayer they made.

zbarnard1 committed:
Add GET User request spec

This commit begins to lay the foundation for our request testing. Down
the line, we'll get the user email from an ID token, but for now it's
simply a header.

zbarnard1 committed:
Add ApiSpecHelpers to parse responses

Adds a spec helper to easily parse the JSON representation of the
records we expect to be returned from the API. Because Timestamps are
returned in the default format using 3 digit precision in iso8601, the
spec helper does the translation for us. Now we don't need to add a
customserializer or a custom matcher for every spec example.

zbarnard1 committed:
Add error handling to GET /user for missing header

The Rack::Request doesn't handle missing headers very well, so when the
required header is missing it throws a KeyError. In this case, we want
to return a response with the proper HTTP status code and a body.

zbarnard1 committed:
Use require_relative app.rb to fix lint error

The two statements are equivalent but one was causing a lint error for
some reason and it was distracting.

zbarnard1 committed:
Let ApiSpecHelpers convert any timestamp

There are a handful of different timestamp attributes, and conditionally
converting them one-by-one is unmaintainable. This commit adds a method
to convert any Time attribute without needing to explicitly check it
before converting.

zbarnard1 committed:
Add .rubocop.yml

I use rubocop locally, and while it's not enabled on this project yet,
it might be nice to move that direction sooner rather than later. Adding
this file to source control lets us use the same config and get the
same results. It would be great if we enabled rubocop to run at some
point before we merge a PR.

It ignores the block length because sometimes specs get long, ignores
the generated schema.rb file because it's not written by us, and
DEFINITELY ignores the documentation requirement because that's
annoying and you should be able to just read the code. We write good
commits so I don't think it's super necessary, but I'm open to a
discussion about it if you think documentation comments would be helpful.

zbarnard1 committed:
Make Resident act_as_list

Residents are linked to a User, and they have a column to track their
position in the list. Using this, we can easily retrieve the next
Resident in the list for the User to pray for.

zbarnard1 committed:
Add acts_as_list gem

I actually thought this was part of ActiveRecord already but it's not.
That said, this is code we don't need to write and I think it solves the
problem really well. Open-source FTW!

zbarnard1 committed:
Add migration for Resident position

Needed to support acts_as_list. Uses a temporary model and an #up and
a #down method due to differences in data needs in either direction.

zbarnard1 commented on app/app.rb:
```diff
@@ -5,32 +5,51 @@
 require 'sinatra/activerecord'
 require 'sinatra/cors'
 require_relative './models'
+register Sinatra::Cors # Add this line.
```
I don't remember where this came from, I think I was trying to get the header to work? This should probably leave unless we need it.

zbarnard1 commented on app/app.rb:
```diff
@@ -5,32 +5,51 @@
-    notation: verse && verse.first.notation,
-    version: verse && verse.first.version
-  }.to_json
+  # get the verse of the day
```
Rubocop auto formatted this, I think I just need to remove it.

ittullos reviewed (changes requested):
Looks great! Could we scope the name validation of the resident to the user?

zbarnard1 committed:
Make Resident name unique for a single User

Multiple Users may upload a Resident with the same
name, so we'll want to scope the uniqueness validation
to the user_id.

ittullos commented on app/models/resident.rb:
```diff
@@ -1,5 +1,15 @@
+
+  def next_resident
+    lower_item || user.residents.limit(1).first
+  end
```
So I'm guessing this gem is just doing all the work for us here and keeping up with our place in the list as we go? On line 13 we're essentially just saying give me the next resident or give me the first one if nil?

zbarnard1 replied:
The gem doesn't really "keep track" of the position, but if we have an instance of a Resident we can navigate down the list using `lower_item`. This returns the next record in the list relative to the instance you called it on.

ittullos commented on spec/support/helpers/api_spec_helpers.rb:
This is my first time seeing API spec helpers. Maybe we can chat a bit about what they do and why they are necessary.

zbarnard1 replied:
Yeah we can dig into it maybe this week, but I'll comment a bit here for now.
The spec helpers are really just a Ruby module. You can include a module in a class and it makes the code in the module available to that class. The idea is that you can separate your re-usable code into modules that can be included where they are needed and it keeps your codebase DRY.
The code in the APISpecHelpers is meant to be used by every request spec, so making it a module makes sharing this code easy. It's not required to use them, but it makes it much easier to parse the response and check it for the data that's important to the test.

ittullos replied:
Awesome! Makes sense!

ittullos reviewed (commented):
This all looks great dude! Was a good refresher for me on a few things too!
